packages/protocol/contracts-0.8/common/EpochManager.sol

```diff
@@ -189,27 +195,36 @@ contract EpochManager is
 
     for (uint i = 0; i < elected.length; i++) {
       address group = getValidators().getValidatorsGroup(elected[i]);
-      if (!processedGroups[group]) {
+      if (!processedGroups[group].processed) {
         epochProcessing.toProcessGroups++;
-        processedGroups[group] = true;
+        uint256 groupScore = getScoreReader().getGroupScore(group);
+        // We need to precompute epoch rewards for each group since computation depends on total active votes for all groups.
+        uint256 epochRewards = getElection().getGroupEpochRewards(
+          group,
+          epochProcessing.totalRewardsVoter,
+          groupScore
+        );
+        processedGroups[group] = ProcessedGroup(true, epochRewards);
       }
     }
 
     require(epochProcessing.toProcessGroups == groups.length, "number of groups does not match");
 
-    for (uint i = 0; i < groups.length; i++) {
+    // since we are adding values it makes sense to start from the end
```

Comment:
I'm assuming the reasoning for this is that values can only increase and thus starting from the end reduces the changes of having to repoint a linked list node?

I am wondering if this changes the assumption for the algorithm that generates the groups, can you think of a case where this could lead to a revert? Can you prove it can't happen?

Reply:
Yep, this should lower the probability of the list to be changed that much. I'm currently looking into algorithm in contractkit and which possibly might be slightly changed - once I have tests I will know more